### PR TITLE
Enable schema queries to be retryable

### PR DESCRIPTION
**Pull request by skipkayhil**
### Motivation / Background

Closes #53425

When `permanent_connection_checkout` was [introduced][1], Active Record was changed to always use `with_connection` and [never][2] `lease_connection`. This causes apps that never call `lease_connection` on their own (or in a gem) during a request to no longer have a connection pinned to the current Thread/Fiber. By checking in/out connections as needed, the number of times a connection must be verified during a single request has significantly increased (more pings = more latency).

### Detail

The good news is that Active Record also now [defers][3] connection verification and can use retryable queries as verification instead of explicit pings.

This commit marks all(?) internal SCHEMA queries as retryable so that they can be used as connection verification. While this likely won't help as much in production for applications using a schema cache dump, it will definitely decrease latency for applications in development and those that do not use a dump in production.

(There are a few places where `query_value{,s}` modifies database state, those were not marked as retryable by this commit)

[1]: https://github.com/rails/rails/commit/4db9f512591b0188f05c0762102a3386b43c1b34
[2]: https://github.com/rails/rails/commit/7c68c5210cbc245d778daa7958cab73bc74f4669
[3]: https://github.com/rails/rails/commit/7fe221d898afe877f282b549bd8bb908ebe1843d

### Additional information

I plan to follow up with more but this PR is already relatively large.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by skipkayhil** [comment]
cc @byroot since you mentioned this before here: https://github.com/rails/rails/pull/52529/commits/9414ed8adde96998a8b8142af633ff2acc9bf10b#r1706748920

File: activerecord/lib/active_record/connection_adapters/postgresql/quoting.rb [hunk 1]
```diff
@@ -193,7 +193,7 @@ def lookup_cast_type_from_column(column) # :nodoc:
 
         private
           def lookup_cast_type(sql_type)
-            super(query_value("SELECT #{quote(sql_type)}::regtype::oid", "SCHEMA").to_i)
+            super(query_value("SELECT #{quote(sql_type)}::regtype::oid", "SCHEMA", allow_retry).to_i)
```
Comment (matthewd) [inline activerecord/lib/active_record/connection_adapters/postgresql/quoting.rb:196]: Typo?

**Review by matthewd — APPROVED** [review]

File: activerecord/lib/active_record/connection_adapters/postgresql/quoting.rb (hunk 1 above)
Comment (skipkayhil) [reply activerecord/lib/active_record/connection_adapters/postgresql/quoting.rb]: Yes thank you, this was super tedious 😓 I double checked all of the others and I think this is the only one

**Comment by skipkayhil** [comment]
Looks like I need an `...` in Postgres, I'm going to make sure tests pass locally before pushing

**Comment by byroot** [comment]
Is the all the `query_values` call sites? Shouldn't we just flip the default value of `allow_retry`?

**Comment by byroot** [comment]
> he number of times a connection must be verified during a single request has significantly increased (more pings = more latency).

Arf, that's not great :/.

**Comment by skipkayhil** [comment]
> Is the all the `query_values` call sites? Shouldn't we just flip the default value of `allow_retry`?

There's a few that appear to change db state, maybe we _could_ retry them but I left them because I wasn't sure

```
active_record/connection_adapters/postgresql_adapter.rb
462:        query_value("SELECT pg_try_advisory_lock(#{lock_id})")
469:        query_value("SELECT pg_advisory_unlock(#{lock_id})")

active_record/connection_adapters/abstract_mysql_adapter.rb
178:        query_value("SELECT GET_LOCK(#{quote(lock_name.to_s)}, #{timeout})") == 1
182:        query_value("SELECT RELEASE_LOCK(#{quote(lock_name.to_s)})") == 1

active_record/connection_adapters/postgresql/schema_statements.rb
323:              query_value("SELECT setval(#{quote(quoted_sequence)}, #{value})", "SCHEMA")
354:            query_value("SELECT setval(#{quote(quoted_sequence)}, #{max_pk || minvalue}, #{max_pk ? true : false})", "SCHEMA")
```

**Comment by byroot** [comment]
> There's a few that appear to change db state,

Hum, the first two are locks, so they can be retried because if the connection is closed, the lock are automatically released.

The last one seem to reset the sequence? It probably shouldn't be a `query_value` call? But something like `exec_query / execute` or something.

**Comment by skipkayhil** [comment]
> > There's a few that appear to change db state,
> 
> Hum, the first two are locks, so they can be retried because if the connection is closed, the lock are automatically released.
> 
> The last one seem to reset the sequence? It probably shouldn't be a `query_value` call? But something like `exec_query / execute` or something.

https://github.com/rails/rails/pull/53425

Opened a new PR with the "retry everything" approach

**Pull request closed without merging**